packages/eslint-plugin/src/rules/keyword-spacing.ts

```diff
@@ -0,0 +1,266 @@
+import baseRule from 'eslint/lib/rules/keyword-spacing';
+import {
+  TSESTree,
+  AST_TOKEN_TYPES,
+} from '@typescript-eslint/experimental-utils';
+import { isTokenOnSameLine, KEYWORDS } from '../util/astUtils';
+import * as util from '../util';
+
+export type Option = Partial<{
+  before: boolean;
+  after: boolean;
+}>;
+export type OverrideOptions = Partial<Record<Keyword, Option>>;
+export type RootOption = Option & { overrides?: OverrideOptions };
+
+export type Options = util.InferOptionsTypeFromRule<typeof baseRule>;
+export type MessageIds = util.InferMessageIdsTypeFromRule<typeof baseRule>;
+
+const PREV_TOKEN = /^[)\]}>]$/u;
+const NEXT_TOKEN = /^(?:[([{<~!]|\+\+?|--?)$/u;
+const TEMPLATE_OPEN_PAREN = '${';
+const TEMPLATE_CLOSE_PAREN = '}';
+const CHECK_TYPE = /^(?:JSXElement|RegularExpression|String|Template)$/u;
+//------------------------------------------------------------------------------
+// Helpers
+//------------------------------------------------------------------------------
+
+/**
+ * Checks whether or not a given token is a "Template" token ends with "${".
+ * @param token A token to check.
+ * @returns `true` if the token is a "Template" token ends with "${".
+ */
+function isOpenParenOfTemplate(token: TSESTree.Token): boolean {
+  return (
+    token.type === AST_TOKEN_TYPES.Template &&
+    token.value.startsWith(TEMPLATE_OPEN_PAREN)
+  );
+}
+
+/**
+ * Checks whether or not a given token is a "Template" token starts with "}".
+ * @param token A token to check.
+ * @returns `true` if the token is a "Template" token starts with "}".
+ */
+function isCloseParenOfTemplate(token: TSESTree.Token): boolean {
+  return (
+    token.type === AST_TOKEN_TYPES.Template &&
+    token.value.endsWith(TEMPLATE_CLOSE_PAREN)
+  );
+}
+
+export default util.createRule<Options, MessageIds>({
+  name: 'keyword-spacing',
+  meta: {
+    type: 'layout',
+    docs: {
+      description: 'Enforce consistent spacing before and after keywords',
+      category: 'Stylistic Issues',
+      recommended: false,
+      extendsBaseRule: true,
+    },
+    fixable: 'whitespace',
+    schema: baseRule.meta.schema,
```

Comment:
Looking over the base rule, I realised it uses the `KEYWORDS` array to build the schema, which means that if we were to add our own keywords this rule would error unless we pulled the schema in ourselves (which wouldn't be the end of the world).

Given that, I think it would be best to have the `KEYWORDS` const in here instead of `astUtils`, to make it internal for just this rule - that way it's less visible and so less likely to be added to by people believing it's a standard utility constant.

This also impacts the typings: while having types as accurate as possible is nice, in this case it doesn't matter as much since we're extending a base rule whos internals we can't reach properly anyway, so the effort required to ensure the types are completely sound is probably going to be far more than the value especially given that we ideally want to make as little changes to this rule as possible, and that it's the schema that powers the validation & such.

In light of this, I personally would say that we don't *need* to copy over the deduplication checking function, since our `KEYWORDS` const should never be different from the one in base rules, and the base rule would error anyway if it changes, but I'm not against it because as you say it's a small number of lines.

------

The tl;dr here is that I think we should definitely keep the `KEYWORDS` constant internal to this rule given how it's used by the base rule.

Reply:
Ok, you got it. So I'll drop the `Keyword[]` type and stick with `string[]`. It will make things simpler.

Reply:
Sounds good. After posting this I realised there are some other keywords that we want to add: 

* `abstract`,
* `asserts`,
* `declare`,
* `extends` (for interfaces),
* `implements`,
* `interface`,
* `infer`,
* `keyof`,
* `module`,
* `namespace`,
* `private`,
* `protected`,
* `public`,
* `readonly`
* `type`,
* `typeof` (this is handled by a different AST node type, and so technically a different operator from the runtime version), 

So we'll want to extend the schema at some point, but I think for now we should just focus on getting the initial rule landed via this PR, as adding new words will have a small diff and so be very easy to do via another PR :)